internal/transport/types.go

```diff
@@ -0,0 +1,50 @@
+package transport
+
+import (
+	"encoding/json"
+	"errors"
+)
+
+var (
+	ErrShuttingDown = errors.New("govim shutting down")
+)
+
+type errProto struct {
+	underlying error
+}
+
+type Transport interface {
+	Start() error
+	Close() error
+	Loaded() chan struct{}
+	Initialized() chan struct{}
+	IsShutdown() chan struct{}
+
+	Read() (id int, messageType string, args []json.RawMessage, err error)
+	Send(callback Callback, callbackType string, params ...interface{}) error
+	SendAndReceive(messageType string, args ...interface{}) (json.RawMessage, error)
+	SendAndReceiveAsync(messageType string, args ...interface{}) (ScheduledCallback, error)
+	SendJSON(p1, p2 interface{}, ps ...interface{})
```

Comment:
Need to remove references to JSON here. Also think we should come up with a better abstraction than `SendAndReceive/SendAndReceiveAsync` - possibly Futures is a good concept to use for the latter (and make the former a helper function that's just sugar around it)